Objective-C/CBLDatabase.mm

```diff
@@ -950,102 +952,127 @@ - (BOOL) saveDocument: (CBLDocument*)document
+        if ($equal(err.domain, CBLErrorDomain) && err.code == CBLErrorConflict)
+            continue;
         
-        return t.commit() || convertError(t.error(), outError);
+        if (outError)
+            *outError = err;
+        
+        return success;
     }
 }
 
-
-- (CBLDocument*) resolveConflictBetweenLocalDoc: (CBLDocument*)localDoc
-                                   andRemoteDoc: (CBLDocument*)remoteDoc {
-    if (remoteDoc.isDeleted)
-        return remoteDoc;
-    else if (localDoc.isDeleted)
-        return localDoc;
-    else if (localDoc.generation > remoteDoc.generation)
-        return localDoc;
-    else if (localDoc.generation < remoteDoc.generation)
-        return remoteDoc;
-    else if ([localDoc.revID compare: remoteDoc.revID] > 0)
-        return localDoc;
-    else
-        return remoteDoc;
-}
-
-
 - (BOOL) saveResolvedDocument: (CBLDocument*)resolvedDoc
                  withLocalDoc: (CBLDocument*)localDoc
                     remoteDoc: (CBLDocument*)remoteDoc
                         error: (NSError**)outError
 {
-    if (resolvedDoc != localDoc)
-        resolvedDoc.database = self;
-    
-    // The remote branch has to win, so that the doc revision history matches the server's.
-    CBLStringBytes winningRevID = remoteDoc.revID;
-    CBLStringBytes losingRevID = localDoc.revID;
-    
-    alloc_slice mergedBody;
-    C4RevisionFlags mergedFlags = 0;
-    if (resolvedDoc != remoteDoc) {
-        // Unless the remote revision is being used as-is, we need a new revision:
-        mergedBody = [resolvedDoc encode: outError];
-        if (!mergedBody)
-            return false;
-        if (resolvedDoc.isDeleted)
-            mergedFlags |= kRevDeleted;
-    }
-    
-    // Tell LiteCore to do the resolution:
-    C4Document *rawDoc = localDoc.c4Doc.rawDoc;
-    C4Error c4err;
-    if (!c4doc_resolveConflict(rawDoc,
-                               winningRevID,
-                               losingRevID,
-                               mergedBody,
-                               mergedFlags,
-                               &c4err)
-        || !c4doc_save(rawDoc, 0, &c4err)) {
-        return convertError(c4err, outError);
+    CBL_LOCK(self) {
+        C4Transaction t(_c4db);
+        if (!t.begin())
+            return convertError(t.error(), outError);
+        
+        if (resolvedDoc != localDoc)
+            resolvedDoc.database = self;
+        
+        // The remote branch has to win, so that the doc revision history matches the server's.
+        CBLStringBytes winningRevID = remoteDoc.revID;
+        CBLStringBytes losingRevID = localDoc.revID;
+        
+        alloc_slice mergedBody;
+        C4RevisionFlags mergedFlags = 0;
+        if (resolvedDoc != remoteDoc) {
+            // Unless the remote revision is being used as-is, we need a new revision:
+            mergedBody = [resolvedDoc encode: outError];
+            if (!mergedBody)
+                return false;
```

Comment:
if resolvedDoc == nil, we don't need to call encode: and we shouldn't skip the check here too.

Reply:
I think the logic here might be something like this (You may need to clean it up):

```
if (!resolvedDoc) {
    mergedFlags |= kRevDeleted;
} else if (resolvedDoc != remoteDoc) {
    mergedBody = [resolvedDoc encode: outError];
    if (!mergedBody)
        return false;
    if (resolvedDoc.isDeleted)
        mergedFlags |= kRevDeleted;
}
```

Reply:
or 

```
if (resolvedDoc != remoteDoc) {
    if (resolvedDoc) {
        mergedBody = [resolvedDoc encode: outError];
        if (!mergedBody)
            return false;
    }

    BOOL isDeleted = !resolvedDoc || resolvedDoc.isDeleted
    if (isDeleted)
        mergedFlags |= kRevDeleted;
}
```

Reply:
is there a chance resolvedDoc is nil, even though the local and remote is not deleted? if any of them is deleted, then the resolvedDoc will not be nil. they will be populated by the logic above 

if that is the case, we will mark the remote as deleted by replacing the local contents right?

Reply:
> is there a chance resolvedDoc is nil, even though the local and remote is not deleted?

Yes. The API doesn't prevent at this moment.

> if any of them is deleted, then the resolvedDoc will not be nil. they will be populated by the logic above

This is correct.

> if that is the case, we will mark the remote as deleted by replacing the local contents right? which is wrong!

I couldn't follow this. Can you explain more?

Reply:
```
        alloc_slice mergedBody;
        C4RevisionFlags mergedFlags = 0;
        // resolvedDoc= nil
        // remote != nil
        // local != nil
        if (resolvedDoc != remoteDoc) {
            BOOL isDeleted = YES;
            if (resolvedDoc) {
                // Unless the remote revision is being used as-is, we need a new revision:
                mergedBody = [resolvedDoc encode: outError];
                if (!mergedBody)
                    return false;
                isDeleted = resolvedDoc.isDeleted;
            }
            
            if (isDeleted) // if resolved doc = nil, meaning the mergedFlags includes kRevDeleted
                mergedFlags |= kRevDeleted;
        }
        // mergedBody  = empty, since resolvedDoc = nil
        // winningRevId is remote.

//  remote will win, and body will be empty. 
```

Reply:
This is correct. Remote should always win. I think empty body is fine because the content of the deleted doc will be empty anyway. If LiteCore doesn't allow nil, we will need to create an empty body.